segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -136,52 +139,187 @@ public void initialize(long epoch) {
+                            isSealed = isSealed(last.getPath());
+                            result = StreamSegmentInformation.builder().name(streamSegmentName).length(last.getLen()).sealed(isSealed).build();
+                            LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName, result);
+                            return result;
+                        });
+        } catch (IOException e) {
+            HDFSExceptionHelpers.throwException(streamSegmentName, e);
+        } catch (RetriesExhaustedException e) {
+            HDFSExceptionHelpers.throwException(streamSegmentName, e.getCause());
+        }
+        return null;
     }
 
     @Override
-    public SegmentProperties create(String streamSegmentName) throws StreamSegmentException {
-        return call(new CreateOperation(streamSegmentName, this.context));
+    public boolean exists(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "exists", streamSegmentName);
+        FileStatus status;
+        try {
+            status = findStatusForSegment(streamSegmentName, false);
+        } catch (IOException e) {
+            return false;
+        }
+        boolean exists = status != null;
+        LoggerHelpers.traceLeave(log, "exists", traceId, streamSegmentName, exists);
+        return exists;
     }
 
-    @Override
-    public SegmentHandle openWrite(String streamSegmentName) throws StreamSegmentException {
-        return call(new OpenWriteOperation(streamSegmentName, this.context));
+    private boolean isSealed(Path path) throws FileNameFormatException {
+        return getEpocFromPath(path) == MAX_EPOC;
+    }
+
+    protected FileSystem openFileSystem(Configuration conf) throws IOException {
+        return FileSystem.get(conf);
     }
 
     @Override
-    public SegmentHandle openRead(String streamSegmentName) throws StreamSegmentException {
-        return call(new OpenReadOperation(streamSegmentName, this.context));
+    public int read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle, offset, length);
+
+        if (offset < 0 || bufferOffset < 0 || length < 0 || buffer.length < bufferOffset + length) {
+            throw new ArrayIndexOutOfBoundsException(String.format(
+                    "Offset (%s) must be non-negative, and bufferOffset (%s) and length (%s) must be valid indices into buffer of size %s.",
+                    offset, bufferOffset, length, buffer.length));
+        }
+
+        Timer timer = new Timer();
+
+        try {
+            return Retry.withExpBackoff(1, 5, MAX_ATTEMPT_COUNT)
+                        .retryingOn(FileNotFoundException.class)
+                        .throwingOn(IOException.class)
+                        .run(() -> {
+                            AtomicInteger totalBytesRead = new AtomicInteger();
```

Comment:
have `readInternal` return an int which is the total bytes read, and get rid of this AtomicInteger

Reply:
Fixed.